src/cli/index.js

```diff
@@ -98,90 +94,55 @@ commander.option('--network-concurrency <number>', 'maximum number of concurrent
 const getDocsInfo = (name) => 'Visit ' + chalk.bold(getDocsLink(name)) + ' for documentation about this command.';
 
-//
+// get command name
+let commandName: string = args.shift() || 'install';
+
 if (commandName === '--help' || commandName === '-h') {
   commandName = 'help';
 }
 
-// if no args or command name looks like a flag then default to `install`
-if (!commandName || commandName[0] === '-') {
-  if (commandName) {
-    args.unshift(commandName);
-  }
-  commandName = 'install';
+if (args.indexOf('--help') >= 0 || args.indexOf('-h') >= 0) {
+  args.unshift(commandName);
+  commandName = 'help';
 }
 
-// aliases: i -> install
-if (commandName && typeof aliases[commandName] === 'string') {
-  const alias = aliases[commandName];
-  command = {
-    run(config: Config, reporter: ConsoleReporter | JSONReporter): Promise<void> {
-      throw new MessageError(`Did you mean \`yarn ${alias}\`?`);
-    },
-  };
+// if no args or command name looks like a flag then set default to `install`
+if (commandName[0] === '-') {
+  args.unshift(commandName);
+  commandName = 'install';
 }
 
-//
-if (commandName === 'help' && args.length) {
-  commandName = camelCase(args.shift());
-  args.push('--help');
+let command;
+const camelised = camelCase(commandName);
+if (camelised) {
+  command = commands[camelised];
 }
 
-//
-invariant(commandName, 'Missing command name');
+// if command is not recognized, then set default to `run`
 if (!command) {
-  const camelised = camelCase(commandName);
-  if (camelised) {
-    command = commands[camelised];
-  }
+  args.unshift(commandName);
+  command = commands.run;
 }
 
-//
 if (command && typeof command.setFlags === 'function') {
   command.setFlags(commander);
 }
 
-if (args.indexOf('--help') >= 0 || args.indexOf('-h') >= 0) {
-  const examples: Array<string> = (command && command.examples) || [];
-  if (examples.length) {
-    commander.on('--help', () => {
-      console.log('  Examples:\n');
-      for (const example of examples) {
-        console.log(`    $ yarn ${example}`);
-      }
-      console.log();
-    });
-  }
-  commander.on('--help', () => console.log('  ' + getDocsInfo(commandName) + '\n'));
-
-  commander.parse(startArgs.concat(args));
-  commander.help();
-  process.exit(1);
-}
-
-args = [commandName].concat(getRcArgs(commandName), args);
-
-if (ARGS_THAT_SHARE_NAMES_WITH_OPTIONS.indexOf(commandName) >= 0 && args[0] === commandName) {
-  args.shift();
-}
-
-commander.parse(startArgs.concat(args));
+commander.parse([
+  ...startArgs,
+  // we use this for https://github.com/tj/commander.js/issues/346, otherwise
+  // it will strip some args that match with any options
+  'cmd',
+  ...getRcArgs(commandName),
+  ...args,
+]);
 commander.args = commander.args.concat(endArgs);
 
-if (command) {
-  commander.args.shift();
-} else {
-  command = commands.run;
-}
-invariant(command, 'missing command');
+// we strip cmd
+commander.args.shift();
```

Comment:
```js
console.assert(commander.args.length >= 1);
console.assert(commander.args[0] === 'cmd');
commander.args.shift();
```

So that the code early crashes instead of doing something else than what the user expected, which might be hard to debug otherwise.